src/Transports.AspNetCore/WebSockets/AsyncMessagePump.cs

```
@@ -0,0 +1,121 @@
+    }
+
+    /// <summary>
+    /// Initializes a new instances with the specified synchronous callback delegate.
+    /// </summary>
+    public AsyncMessagePump(Action<T> callback)
+    {
+        if (callback == null)
+            throw new ArgumentNullException(nameof(callback));
+        _callback = message =>
+        {
+            callback(message);
+            return Task.CompletedTask;
+        };
+    }
+
+    /// <summary>
+    /// Posts the specified message to the message queue.
+    /// </summary>
+    public void Post(T message)
+        => Post(new ValueTask<T>(message));
+
+    /// <summary>
+    /// Posts the result of an asynchronous operation to the message queue.
+    /// </summary>
+    public void Post(ValueTask<T> messageTask)
+    {
+        bool attach = false;
+        lock (_queue)
+        {
+            _queue.Enqueue(messageTask);
+            attach = _queue.Count == 1;
+        }
+
+        if (attach)
+        {
+            _ = CompleteAsync();
+        }
+    }
+
+    /// <summary>
+    /// Processes message in the queue until it is empty.
+    /// </summary>
+    private async Task CompleteAsync()
+    {
+        // grab the message at the start of the queue, but don't remove it from the queue
+        ValueTask<T> messageTask;
+        lock (_queue)
+        {
+            // should always successfully peek from the queue here
+#pragma warning disable CA2012 // Use ValueTasks correctly
+            messageTask = _queue.Peek();
+#pragma warning restore CA2012 // Use ValueTasks correctly
+        }
+        while (true)
+        {
+            // process the message
+            try
+            {
+                var message = await messageTask.ConfigureAwait(false);
+                await _callback(message).ConfigureAwait(false);
+            }
+            catch (Exception ex)
+            {
+                try
+                {
+                    await HandleErrorAsync(ex);
+                }
+                catch { }
```

I would copy/paste xml comments from `ObservableExtensions` here to explain empty catch block.

---

```suggestion
                catch
                {
                    // if an error is unhandled, execution of this method will terminate, and
                    // no further events in the queue will be processed, and Queue will not
                    // start another ProcessAllEventsInQueueAsync because the queue is not empty
                 }
```